### PR TITLE
V1 of the A2 boilerplate.

### PR DESCRIPTION
**Pull request by lakinwecker**

### CODE REVIEW

File: 453-skeleton/main.cpp [hunk 1]
```diff
@@ -55,37 +75,96 @@ int main() {
...
-	cpuGeom.verts.push_back(glm::vec3(0.f, 0.5f, 0.f));
-
-	// colours (these should be in linear space)
-	cpuGeom.cols.push_back(glm::vec3(1.f, 0.f, 0.f));
-	cpuGeom.cols.push_back(glm::vec3(0.f, 1.f, 0.f));
-	cpuGeom.cols.push_back(glm::vec3(0.f, 0.f, 1.f));
-
-	gpuGeom.setVerts(cpuGeom.verts);
-	gpuGeom.setCols(cpuGeom.cols);
+	// GL_NEAREST looks a bit better for low-res pixel art than GL_LINEAR.
```
Comment (lakinwecker) [inline 453-skeleton/main.cpp:112]: I feel like, for this example, we should move the ship coordinate generation into a separate function. Not a big deal, but my feelings anyways.

File: 453-skeleton/GLHandles.cpp [hunk 2]
```diff
@@ -145,3 +145,53 @@ GLuint VertexBufferHandle::value() const {
...
+}
+
+TextureHandle& TextureHandle::operator=(TextureHandle&& other) noexcept {
+	std::swap(textureID, other.textureID);
+	return *this;
+}
+
+
+TextureHandle::~TextureHandle() {
+	if (TextureHandle::lastBoundTexture == textureID) {
```
Comment (lakinwecker) [inline 453-skeleton/GLHandles.cpp]: Is this necessary?  

I feel like it's not.  Let's chat about this.

**Comment by lakinwecker** [comment]
I don't feel that the static is necessary. It also won't even work properly in the context of a multi-threaded program, although neither does OpenGL so maybe that's not a concern.

If you want to use RAII to do texture binding, then you should create a new class `BoundTexture`, which takes a texture as it's only argument, binds in the constructor, and then unbinds in the destructor, then in your code, you use:
```
{ 
    auto tex = BoundTexture(myTexture);
    ....
    /// draw calls
} /// Unbind is called here.
```


But I'm not even sure that's necessary for this

File: 453-skeleton/Texture.cpp [hunk 3]
```diff
@@ -0,0 +1,88 @@
+#include "Texture.h"
+#define STB_IMAGE_IMPLEMENTATION
+#include <stb/stb_image.h>
+
+#include <iostream>
+
+Texture::Texture(std::string path, GLint interpolation)
+	: textureID(), path(path), interpolation(interpolation)
```
Comment (lakinwecker) [inline 453-skeleton/Texture.cpp:8]: These need to be initialized in the order they are declared in the .h file.

File: 453-skeleton/Texture.cpp [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+		if (CheckGLErrors((std::string("Loading texture: ") + path).c_str())) {
+			throw std::runtime_error("Texture failed to load!");
+		}
+	}
+	else {
+		throw std::runtime_error("Failed to read texture data from file!");
+	}
+}
+
+bool Texture::CheckGLErrors(const char* errorLocation)
```
Comment (lakinwecker) [inline 453-skeleton/Texture.cpp]: We already have the GLDebug output, I would assume that catches all of these errors already. I don't mind leaving this here if you have any concerns that it won't catch it.

Comment (cm-1) [reply 453-skeleton/Texture.cpp]: Nah, I just copied that function in from the boilerplate from when I took the course. I'll remove it.

File: 453-skeleton/Texture.cpp (hunk 3 above)
Comment (cm-1) [reply 453-skeleton/Texture.cpp:8]: Fixed.

**Pull request merged**